### ISSUE TITLE
BUG: SSH Keys in E2Es 

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#69373]
**Is this a BUG REPORT**:
There are numerous e2e tests which require SSH keys to be present, otherwise they are skipped.

**What happened**:
It's a weird behavior validation that is not API-driven or agnostic, a lot of those tests get skipped b/c of restrictions on how the e2es run. 

**What you expected to happen**:
e2es should be behavioral driven.  If we can not validate via the API then there is a gap.  A lot of these checks were added early in k8s life and should be removed. 

/cc @kubernetes/sig-testing @kubernetes/cncf-conformance-wg
/kind bug 
/sig testing

**Comment by stevekuznetsov** [comment]
Can we list the tests here to track progress on the issue?

**Comment by johnSchnake** [comment]
As requested above; I'll do an audit and create a list of the affected tests. We can discuss further at that point.

**Comment by johnSchnake** [comment]
Files with framework.SkipUnlessSSHKeyPresent()
 - test/e2e/network/service.go
 - test/e2e/node/crictl.go
 - test/e2e/node/node_problem_detector.go
 - test/e2e/node/ssh.go
 - test/e2e/framework/util.go
 - test/e2e/storage/nfs_persistent_volume-disruptive.go
 - test/e2e/storage/flexvolume_mounted_volume_resize.go
 - test/e2e/storage/flexvolume_online_resize.go
 - test/e2e/storage/detach_mounted.go
 - test/e2e/storage/flexvolume.go

Tests:
Services 
 - should be able to up and down services
 - should implement service.kubernetes.io/service-proxy-name

crictl (beforeEach but only one test)
 - should be able to run crictl on the node

NodeProblemDetector [DisabledForLargeClusters] (beforeEach)
 - should run without error

SSH
 - should SSH to all nodes and run commands
 
Detaching volumes (BeforeEach)
 - should not work when mount is in progress [Slow]

Mounted flexvolume expand[Slow] (BeforeEach)
 - Should verify mounted flex volumes can be resized

Mounted flexvolume volume expand [Slow] [Feature:ExpandInUsePersistentVolumes] (BeforeEach)
 - should be resizable when mounted

Flexvolumes
 - should be mountable when non-attachable
 - should be mountable when attachable

NFSPersistentVolumes[Disruptive][Flaky]
 - when kube-controller-manager restarts
   - should delete a bound PVC from a clientPod, restart the kube-control-manager, and ensure the kube-controller-manager does not crash

For some of those tests, the reason they need SSH is to use the `installFlex` method.

However, SSH is also used in methods like `RestartControllerManager` and (frustratingly) not everything bubbles this "skip" requirement up. Utilities like this are used in tests that don't gate for SSH at all (i.e. "should reconcile LB health check interval [Slow][Serial]"). This makes it very difficult to try and pull out any/all use of SSH.

In addition, 22 files contain calls to SSH directly and there are a few other SSH methods which may be called directly in tests or in helpers which are then used in (potentially) multiple other tests.

 - test/e2e/apimachinery/etcd_failure.go
 - test/e2e/framework/get-kubemark-resource-usage.go
 - test/e2e/framework/log_size_monitoring.go
 - test/e2e/framework/metrics_util.go
 - test/e2e/framework/profile_gatherer.go
 - test/e2e/framework/service_util.go
 - test/e2e/framework/ssh.go
 - test/e2e/framework/util.go
 - test/e2e/network/service.go
 - test/e2e/node/crictl.go
 - test/e2e/node/node_problem_detector.go
 - test/e2e/node/ssh.go
 - test/e2e/storage/flexvolume.go
 - test/e2e/storage/utils/utils.go
 - test/e2e/storage/vsphere/vsphere_utils.go
 - test/e2e_node/remote/cadvisor_e2e.go
 - test/e2e_node/remote/node_conformance.go
 - test/e2e_node/remote/node_e2e.go
 - test/e2e_node/remote/remote.go
 - test/e2e_node/remote/ssh.go
 - test/e2e_node/remote/utils.go
 - test/e2e_node/runner/remote/run_remote.go

**Comment by johnSchnake** [comment]
No actions specifically taken on this; it has now been split up which makes recognizing the dependency more simple but each situation would still have to find another way to test what it is trying to check.

**Comment by johnSchnake** [comment]
TODO: 
 [ ] Make a firm list of exact tests which require SSH, noting if they are conformance tests or not
 [ ] Each of those tests should have the skip-unless-ssh method
 [ ] Following ^, this should prevent those tests from being conformance tests (since they have skip)
 [ ] The conformance tests need modified to not use SSH

The remaining tests can be considered by testing-commons; but keep in mind that having the SSH behavior required would be a blocker for the test ever becoming a conformance test.

**Comment by johnSchnake** [comment]
So this is a frustratingly hard problem due to how large the list of functions which are tied to ssh is.

I made an ast tool to try and assist with this and have iterated on that quite a bit today. I will spend a bit of time validating the results I've gotten and turning it into a reasonable looking list.

TIL: It is really hard to do something like simply list all the callers of a function when accounting for all  cases. Functions that wrap functions that wrap functions are frustrating...

**Comment by johnSchnake** [comment]
For conformance tests, it may be easier to do something like check the call stack when ssh'ing and ensuring we aren't in a conformance test and erroring if we are. Otherwise, I don't know how we'd enforce this.

**Comment by johnSchnake** [comment]
Below is a link to the full set of tests that I found linked to SSH behavior.

TL;DR;

Main suite: 32 tests as well as the entire `framework.AfterEach` when collecting metrics.

All node conformance: 150 tests
 - multiple methods like setupCNI use SSH when running each test

Scalability - 12 density tests
Scalability - 10 load capacity tests

2 unit tests related volume mounts.

Full dump here: https://docs.google.com/document/d/1-DYYw-QoOC3lztfCmOPAbd52K-iM_uw1MU_NjkZ0qdg/edit?usp=sharing

**Comment by johnSchnake** [comment]
- There are currently no `Conformance` tests that use the SSH features that I was able to find (except the `NodeConformance`.
 - Seems reasonable to add skips for the 32 other tests; thats not too bad
 - Do we care the scalability tests use SSH? It is in a different suite entirely.

Update/correction to the above: The `framework.AfterEach` calls a method `StopAndSummarize` but I think in that context it is a `ContainerResourceGatherer` object calling it and not the `EtcdMetricsCollector`. As a result, the main `framework.AfterEach` does not rely on SSH as far as I can tell. The `EtcdMetricsCollector` is used only in other scalability tests as far as I can tell (which already have SSH requirements too)
